playground/flappy-bird.ts

```diff
@@ -0,0 +1,798 @@
+    if (leftPosition.x < -groundWid) {
+      // The invisible ground can be moved to the right
+      leftPosition.x = rightPosition.x + groundWid;
+    }
+    leftTransform.position = leftPosition;
+    rightTransform.position = rightPosition;
+    // Swap
+    var temp = this.leftGround;
+    this.leftGround = this.rightGround;
+    this.rightGround = temp;
+  }
+}
+
+class GameCtrl extends Script {
+  private _gameState: EnumGameState;
+  private _tempViewportVec2 = new Vector2();
+  private _tempRay = new Ray();
+  onAwake() {
+    const { entity } = this;
+    engine.on(GameEvent.reStartGame, () => {
+      this.gameState = EnumGameState.Idel;
+    });
+
+    engine.on(GameEvent.gameOver, () => {
+      this.gameState = EnumGameState.End;
+    });
+
+    const webCanvas = engine.canvas._webCanvas;
+    const camera = entity.findByName("camera").getComponent(Camera);
+    const viewport = camera.viewport;
+    const tempViewportVec2 = this._tempViewportVec2;
+    const tempRay = this._tempRay;
+    const onTouchEnd = (touchEvt) => {
+      if (!touchEvt) {
+        return;
+      }
+      switch (this._gameState) {
+        case EnumGameState.End:
+          const { changedTouches = [], target } = touchEvt;
+          const { left = 0, top = 0 } = target && target.getBoundingClientRect();
+          if (changedTouches.length > 0) {
+            const mPos = changedTouches[0];
+            tempViewportVec2.setValue(
+              ((mPos.clientX - left) / webCanvas.clientWidth - viewport.x) / viewport.z,
+              ((mPos.clientY - top) / webCanvas.clientHeight - viewport.y) / viewport.w
+            );
+            camera.viewportPointToRay(tempViewportVec2, tempRay);
+            engine.dispatch(GameEvent.checkHitGui, tempRay);
+          }
+          if (touchEvt) break;
+        default:
+          break;
+      }
+    };
+    const onMouseUp = (touchEvt) => {
+      if (!touchEvt) {
+        return;
+      }
+      switch (this._gameState) {
+        case EnumGameState.End:
+          const { pageX, pageY, target } = touchEvt;
+          const { offsetLeft = 0, offsetTop = 0 } = target;
+          tempViewportVec2.setValue(
+            ((pageX - offsetLeft) / webCanvas.clientWidth - viewport.x) / viewport.z,
+            ((pageY - offsetTop) / webCanvas.clientHeight - viewport.y) / viewport.w
+          );
+          camera.viewportPointToRay(tempViewportVec2, tempRay);
+          engine.dispatch(GameEvent.checkHitGui, tempRay);
+        default:
+          break;
+      }
+    };
+    const onMouseDown = () => {
+      switch (this._gameState) {
+        case EnumGameState.Idel:
+          this.gameState = EnumGameState.Start;
+          engine.dispatch(GameEvent.fly);
+          break;
+        case EnumGameState.Start:
+          engine.dispatch(GameEvent.fly);
+          break;
+        default:
+          break;
+      }
+    };
+    //Monitor mouse events
+    webCanvas.addEventListener(CanvasEvent.mouseup, onMouseUp);
```

Comment:
Abstract the codes to a simple lib such as *@oasis-engine/gesture* is better. We can add more features to the lib in the feature.

Reply:
Go back to abstract afterwards